charms/jupyter-ui/src/charm.py

```diff
@@ -25,139 +39,190 @@ def __init__(self, msg, status_type=None):
-                                "files": [
-                                    {
-                                        "path": name,
-                                        "content": content,
-                                    }
-                                    for name, content in yaml.safe_load(
-                                        Path("src/logos-configmap.yaml").read_text()
-                                    )["data"].items()
-                                ],
-                            },
-                        ],
-                    }
-                ],
             },
+        }
+
+        return Layer(layer_config)
+
+    def _update_layer(self) -> None:
+        """Update the Pebble configuration layer (if changed)."""
+        current_layer = self.container.get_plan()
+        new_layer = self._jupyter_ui_layer
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self.container.add_layer(self._container_name, new_layer, combine=True)
+            try:
+                self.logger.info("Pebble plan updated with new configuration, replaning")
+                self.container.replan()
+            except ChangeError:
+                raise ErrorWithStatus("Failed to replan", BlockedStatus)
+
+    def _upload_files_to_container(self):
+        """Upload required files to container."""
+        self.container.push(
+            "/etc/config/spawner_ui_config.yaml",
+            "spawner_ui_config.yaml",
+            make_dirs=True,
         )
-        self.model.unit.status = ActiveStatus()
+        for file_name, file_content in yaml.safe_load(
+            Path("src/logos-configmap.yaml").read_text()
+        )["data"].items():
+            logo_file = "/src/apps/default/static/assets/logos/" + file_name
+            self.container.push(
+                logo_file,
+                file_content,
+                make_dirs=True,
+            )
+
+    def _deploy_k8s_resources(self) -> None:
+        """Deploys K8S resources."""
+        try:
+            self.unit.status = MaintenanceStatus("Creating K8S resources")
+            self.k8s_resource_handler.apply()
+        except ApiError:
+            raise ErrorWithStatus("K8S resources creation failed", BlockedStatus)
+        self.model.unit.status = MaintenanceStatus("K8S resources created")
+
+    def _on_install(self, _):
+        """Perform installation only actions."""
+        try:
+            self._deploy_k8s_resources()
+        except CheckFailed as err:
+            self.model.unit.status = err.status
+        return
+
+    def _on_pebble_ready(self, _):
+        """Configure started container."""
+        if not self._is_container_ready():
+            return
+
+        # upload files to container
+        self._upload_files_to_container()
+
+        # proceed with other actions
+        self.main(_)
```

Comment:
We talked about this a bit, but highlighting since it is easy for someone else to miss in a review and they might have an opinion on it.

This works I think, but it still feels funny.  We have two possible event sequences:

Install first:
a. install: 
  * deploys the k8s resources (`deploy_k8s_resources()`)
  * does not "wake" the rest of the charm (does not call `main`).  This is because the container isn't up yet, so `main()` would fail right now
b. pebble-ready: 
  * our container is now alive, so we transfer files to it and then trigger `main()` to wake up the charm
  * `main()` calls `deploy_k8s_resources()` a second time.  Probably nothing has changed and this is a no-op anyway

Pebble-ready first:
1. pebble-ready: 
  * container is alive, so we transfer files and trigger `main()`
  * `main()` calls `deploy_k8s_resources()`, deploying all the k8s resources we need
2. install: 
  * installs k8s resources a second time.  If settings haven't changed since before, this is effectively a no-op.  **But, if something has changed** (config change?  can that happen after `pebble-ready` and before `install`?) then we will update the k8s resources
  * we do not call `main` again, so if we did change our k8s resources our main (eg: our charm) will not be in sync with those resources.  This probably isn't bad here, but could be in some charms

I'm not sure what to think.  This feels like something that is probably ok here, but there might be a better way.  And in a few of our charms, this pattern might result in a subtle intermittent error that's hard to figure out.  Thoughts?

Reply:
We might be saved here by how juju works.  I don't think it is possible to have a `config-changed` between `pebble-ready` and `install`, but I'm not sure

Reply:
I don't think `pebble-ready` can happen before `install`:
https://juju.is/docs/sdk/container-name-pebble-ready-event#heading--emission-sequence

Reply:
The <container name>-pebble-ready event is emitted once the Pebble sidecar container has started and a socket is available.